In [1]:
# The first step of using numpy is to tell python to import it
import numpy as np


### 2.1 NumPy Arrays<a href="#2.1-NumPy-Arrays" class="anchor-link">¶</a>

**NumPy Array**

-   An array is a data structure that stores values of same data type.
-   While python lists can contain values corresponding to different
    data types, arrays in python can only contain values corresponding
    to the same data type.
-   However python lists fail to deliver the performance required while
    computing large sets of numerical data. To address this issue we use
    NumPy arrays.
-   We can create NumPy arrays by converting a list to an array.



In [2]:
# defining a list of different car companies or string elements
arr_str = ['Mercedes', 'BMW', 'Audi', 'Ferrari', 'Tesla']

# defining a list of number of cylinders in car or numerical elements
arr_num = [5, 4, 6, 7, 3]

In [3]:
# connverting the list arr_str to a NumPy array
np_arr_str = np.array(arr_str)

In [4]:
# connverting the list arr_num to a NumPy array
np_arr_num = np.array(arr_num)

In [6]:
# checking the output
print('Numpy Array (arr_str): ',np_arr_str)
print('Numpy Array (arr_num): ',np_arr_num)

Numpy Array (arr_str):  ['Mercedes' 'BMW' 'Audi' 'Ferrari' 'Tesla']
Numpy Array (arr_num):  [5 4 6 7 3]


The resuts look similar to a list but arr_str and arr_num have been
converted to NumPy arrays. Let's check the data type to confirm this.

In [7]:
# printing the data type of lists
print('Data type of arr_str: ',type(arr_str))
print('Data type of arr_num: ',type(arr_num))

# printing the data type after conversion of lists to array
print('Data type of np_arr_str: ',type(np_arr_str))
print('Data type of np_arr_num: ',type(np_arr_num))

Data type of arr_str:  <class 'list'>
Data type of arr_num:  <class 'list'>
Data type of np_arr_str:  <class 'numpy.ndarray'>
Data type of np_arr_num:  <class 'numpy.ndarray'>


-   The above output confirms that both the lists were successfully
    converted to arrays

**NumPy Matrix**

-   A matrix is a two-dimensional data structure where elements are
    arranged into rows and columns.
-   A matrix can be created by using list of lists

In [8]:
# let's say we have information of different number of cylinders in a car and we want to display them in a matrix format
matrix = np.array([[1,2,1],[4,5,9],[1,8,9]])
print(matrix)

[[1 2 1]
 [4 5 9]
 [1 8 9]]


In [9]:
print('Data type of matrix: ',type(matrix))

Data type of matrix:  <class 'numpy.ndarray'>


-   We see that all the NumPy objects have data type as ndarray

### 2.2 NumPy Functions<a href="#2.2-NumPy-Functions" class="anchor-link">¶</a>

**There are different ways to create NumPy arrays using the functions
available in NumPy library**

**Using np.arange() function**

-   The np.arange() function returns an array with evenly spaced
    elements as per the interval. The interval mentioned is half-opened
    i.e. start is included but stop is excluded.
-   It has the following paramaters:
    -   start : start of interval range. By default start = 0
    -   stop : end of interval range
    -   step : step size of interval. By default step size = 1

In [10]:
arr2  = np.arange(start = 0, stop = 10) # 10 will be excluded from the output
print(arr2)

[0 1 2 3 4 5 6 7 8 9]


In [11]:
# or
arr2  = np.arange(0,10) 
print(arr2)

[0 1 2 3 4 5 6 7 8 9]


In [12]:
# adding a step size of 5 to create an array
arr3  = np.arange(start = 0, stop = 20, step = 5)
arr3

array([ 0,  5, 10, 15])

**Using np.linspace() function**

-   The np.linspace() function returns numbers which are evenly
    distributed with respect to interval. Here the start and stop both
    are included.

\*It has the following parameters:

-   start: start of interval range. By default start = 0
-   stop: end of interval range
-   num : No. of samples to generate. By default num = 50

In [13]:
matrix2 = np.linspace(0,5) # by default 50 evenly spaced values will be generated between 0 and 5
matrix2

array([0.        , 0.10204082, 0.20408163, 0.30612245, 0.40816327,
       0.51020408, 0.6122449 , 0.71428571, 0.81632653, 0.91836735,
       1.02040816, 1.12244898, 1.2244898 , 1.32653061, 1.42857143,
       1.53061224, 1.63265306, 1.73469388, 1.83673469, 1.93877551,
       2.04081633, 2.14285714, 2.24489796, 2.34693878, 2.44897959,
       2.55102041, 2.65306122, 2.75510204, 2.85714286, 2.95918367,
       3.06122449, 3.16326531, 3.26530612, 3.36734694, 3.46938776,
       3.57142857, 3.67346939, 3.7755102 , 3.87755102, 3.97959184,
       4.08163265, 4.18367347, 4.28571429, 4.3877551 , 4.48979592,
       4.59183673, 4.69387755, 4.79591837, 4.89795918, 5.        ])

**How are these values getting generated?**

The step size or the difference between each element will be decided by
the following formula:

**(stop - start) / (total elements - 1)**

So, in this case: (5 - 0) / 49 = 0.10204082

The first value will be 0.10204082, the second value will be
0.10204082 + 0.10204082, the third value will be 0.10204082 + 0.10204082
+0.10204082, and so on.

In [14]:
# generating 10 evenly spaced values between 10 and 20
matrix3 = np.linspace(10,20,10)
matrix3

array([10.        , 11.11111111, 12.22222222, 13.33333333, 14.44444444,
       15.55555556, 16.66666667, 17.77777778, 18.88888889, 20.        ])

**Similarly we can create matrices using the functions available in
NumPy library**

**Using np.zeros()**

-   The np.zeros() is a function for creating a matrix and performing
    matrix operations in NumPy.
-   It returns a matrix filled with zeros of the given shape.
-   It has the following parameters:
    -   shape : Number of rows and columns in the output matrix.
    -   dtype: data type of the elements in the matrix, by default the
        value is set to `float`.

In [15]:
matrix4 = np.zeros([3,5])
matrix4

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

**Using np.ones()**

-   The np.ones() is another function for creating a matrix and
    performing matrix operations in NumPy.
-   It returns a matrix of given shape and type, filled with ones.
-   It has the following parameters:
    -   shape : Number of rows and columns in the output matrix.
    -   dtype: data type of the elements in the matrix, by default the
        value is set to `float`.

In [16]:
matrix5 = np.ones([3,5])
matrix5

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

**Using np.eye()**

-   The np.eye() is a function for creating a matrix and performing
    matrix operations in NumPy.
-   It returns a matrix with ones on the diagonal and zeros elsewhere.
-   It has the following parameters:
    -   n: Number of rows and columns in the output matrix
    -   dtype: data type of the elements in the matrix, by default the
        value is set to `float`.

In [17]:
matrix6 = np.eye(5)
matrix6

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

**We can also convert a one dimension array to a matrix. This can be
done by using the np.reshape() function.**

-   The shape of an array basically tells the number of elements and
    dimensions of the array. Reshaping a Numpy array simply means
    changing the shape of the given array.
-   By reshaping an array we can add or remove dimensions or change
    number of elements in each dimension.
-   In order to reshape a NumPy array, we use the reshape method with
    the given array.
-   **Syntax:** array.reshape(shape)
    -   shape: a tuple given as input, the values in tuple will be the
        new shape of the array.

In [18]:
# defining an array with values 0 to 9
arr4 = np.arange(0,10) 
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [19]:
# reshaping the array arr4 to a 2 x 5 matrix
arr4_reshaped = arr4.reshape((2,5))
arr4_reshaped

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [20]:
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [21]:
# reshaping the array arr4 to a 2 x 6 matrix
arr4.reshape((2,6))

ValueError: cannot reshape array of size 10 into shape (2,6)

-   This did not work because we have 10 elements which we are trying to
    fit in a 2 X 6 shape which will require 12 elements.

**NumPy can also perform a large number of different mathematical
operations and it provides different functions to do so.**

NumPy provides:

1.  Trigonometric functions
2.  Exponents and Logarithmic functions
3.  Functions for arithmetic operations between arrays and matrices

**Trigonometric functions**

In [22]:
print('Sine Function:',np.sin(4))
print('Cosine Function:',np.cos(4))
print('Tan Function',np.tan(4))

Sine Function: -0.7568024953079282
Cosine Function: -0.6536436208636119
Tan Function 1.1578212823495777


**Exponents and Logarithmic functions**

-   Exponents

In [23]:
np.exp(2)

7.38905609893065

In [25]:
arr5 = np.array([2,4,6])
np.exp(arr5)

array([  7.3890561 ,  54.59815003, 403.42879349])

-   Logarithms

In [26]:
# by default NumPy takes the base of log as e
np.log(2)

0.6931471805599453

In [27]:
np.log(arr5)

array([0.69314718, 1.38629436, 1.79175947])

In [28]:
# log with base 10
np.log10(8)

0.9030899869919435

**Arithmetic Operations on arrays**

In [29]:
# arithmetic on lists

l1 = [1,2,3]
l2 = [4,5,6]
print(l1+l2)
# this does not behave as you would expect!

[1, 2, 3, 4, 5, 6]


In [30]:
# we can +-*/ arrays together

# defining two arrays
arr7 = np.arange(1,6)
print('arr7:', arr7)

arr8 = np.arange(3,8)
print('arr8:', arr8)

arr7: [1 2 3 4 5]
arr8: [3 4 5 6 7]


In [31]:
print('Addition: ',arr7+arr8)
print('Subtraction: ',arr8-arr7)
print('Multiplication:' , arr7*arr8)
print('Division:', arr7/arr8)
print('Inverse:', 1/arr7)
print('Powers:', arr7**arr8) # in python, powers are achieved using **, NOT ^!!! ^ does something completely different!

Addition:  [ 4  6  8 10 12]
Subtraction:  [2 2 2 2 2]
Multiplication: [ 3  8 15 24 35]
Division: [0.33333333 0.5        0.6        0.66666667 0.71428571]
Inverse: [1.         0.5        0.33333333 0.25       0.2       ]
Powers: [    1    16   243  4096 78125]


**Operations on Matrices**

In [32]:
matrix7 = np.arange(1,10).reshape(3,3)
print(matrix7)

matrix8 = np.eye(3)
print(matrix8)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [33]:
print('Addition: \n', matrix7+matrix8)
print('Subtraction: \n ', matrix7-matrix8)
print('Multiplication: \n', matrix7*matrix8)
print('Division: \n', matrix7/matrix8)

Addition: 
 [[ 2.  2.  3.]
 [ 4.  6.  6.]
 [ 7.  8. 10.]]
Subtraction: 
  [[0. 2. 3.]
 [4. 4. 6.]
 [7. 8. 8.]]
Multiplication: 
 [[1. 0. 0.]
 [0. 5. 0.]
 [0. 0. 9.]]
Division: 
 [[ 1. inf inf]
 [inf  5. inf]
 [inf inf  9.]]


C:\Users\acer\AppData\Local\Temp\ipykernel_20192\2604038799.py:4: RuntimeWarning: divide by zero encountered in divide
  print('Division: \n', matrix7/matrix8)


-   RuntimeWarning: Errors which occur during program
    execution(run-time) after successful compilation are called run-time
    errors.
-   One of the most common run-time error is division by zero also known
    as Division error.
-   Due to division by zero error, we are getting inf (infinity) values
    because 1/0 is not a defined operation.

**Linear algebra matrix multiplication**

In [34]:
matrix9 = np.arange(1,10).reshape(3,3)
print('First Matrix: \n',matrix9)

matrix10 = np.arange(11,20).reshape(3,3)
print('Second Matrix: \n',matrix10)
print('')
# taking linear algebra matrix multiplication (some may have heard this called the dot product)
print('Multiplication: \n', matrix9 @ matrix10)

First Matrix: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Second Matrix: 
 [[11 12 13]
 [14 15 16]
 [17 18 19]]

Multiplication: 
 [[ 90  96 102]
 [216 231 246]
 [342 366 390]]


**Transpose of a matrix**

In [35]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [36]:
# taking transpose of matrix
np.transpose(matrix9)

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [37]:
# another way of taking a transpose
matrix9.T

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

**Function to find minimum and maximum values**

In [38]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [39]:
print('Minimum value: ',np.min(matrix9))

Minimum value:  1


In [40]:
print('Maximum value: ',np.max(matrix9))

Maximum value:  9


**Function to generate random samples**

**Using np.random.rand function**

-   The np.random.rand returns a random NumPy array whose element(s) are
    drawn randomly from the uniform distribution over \[0,1). (including
    0 but excluding 1).
-   **Syntax** - np.random.rand(d0,d1)
    -   d0,d1 – It represents the dimension of the required array given
        as int, where d1 is optional.

In [41]:
# Generating random values in an array
rand_mat = np.random.rand(5)
print(rand_mat)

[0.80551497 0.64690159 0.17244782 0.83167654 0.92109335]


In [42]:
# * Generating random values in a matrix
rand_mat = np.random.rand(5,5) # uniform random variable
print(rand_mat)

[[0.3824405  0.12945818 0.86296332 0.68122915 0.78301093]
 [0.25507064 0.54851148 0.91919841 0.01953608 0.54438823]
 [0.05442505 0.92588402 0.07091178 0.31137578 0.55992409]
 [0.54936662 0.47034121 0.22434761 0.09906725 0.9378072 ]
 [0.29242805 0.47472319 0.66199508 0.18600803 0.84116451]]


**Using np.random.randn function**

-   The np.random.randn returns a random numpy array whose sample(s) are
    drawn randomly from the standard normal distribution (Mean as 0 and
    standard deviation as 1)

-   **Syntax** - np.random.randn(d0,d1)

    -   d0,d1 – It represents the dimension of the output, where d1 is
        optional.

In [46]:
# Generating random values in an array
rand_mat2 = np.random.randn(5) 
print(rand_mat2)

[-0.33746134 -1.10316022 -1.03029278 -1.09823378 -1.70360759]


In [49]:
# Generating random values in a matrix
rand_mat2 = np.random.randn(5,5) 
print(rand_mat2)

[[-0.00253289 -0.21348875 -0.19114209 -0.23689337  0.73035447]
 [-0.91926642 -0.04236675  0.79709449 -1.40645231 -0.78280153]
 [-0.94362946 -0.1647715   0.21638714  1.75021631  0.25519348]
 [-0.68742509 -0.59891315  0.39795513 -0.74337474  1.09198669]
 [ 0.15960987 -0.49583804 -1.1237828   2.07370523  1.06382067]]


In [50]:
# Let's check the mean and standard deviation of rand_mat2
print('Mean:',np.mean(rand_mat2))
print('Standard Deviation:',np.std(rand_mat2))

Mean: -0.0006542152448686522
Standard Deviation: 0.861124204431122


-   We observe that the mean is very close to 0 and standard deviation
    is very close to 1.

**Using np.random.randint function**

-   The np.random.randint returns a random numpy array whose element(s)
    are drawn randomly from low (inclusive) to the high (exclusive)
    range.

-   **Syntax** - np.random.randint(low, high, size)

    -   low – It represents the lowest inclusive bound of the
        distribution from where the sample can be drawn.
    -   high – It represents the upper exclusive bound of the
        distribution from where the sample can be drawn.
    -   size – It represents the shape of the output.

In [58]:
# Generating random values in an array
rand_mat3 = np.random.randint(1,5,10)
print(rand_mat3)

[3 2 2 4 3 1 1 1 4 1]


In [62]:
# Generating random values in a matrix
rand_mat3 = np.random.randint(1,10,[5,5])
print(rand_mat3)

[[7 6 1 1 6]
 [8 1 8 5 8]
 [2 8 8 9 9]
 [4 4 6 7 1]
 [7 2 3 2 3]]


### 2.3 Accessing the entries of a Numpy Array<a href="#2.3-Accessing-the-entries-of-a-Numpy-Array" class="anchor-link">¶</a>

In [63]:
# let's generate an array with 10 random values
rand_arr = np.random.randn(10)
print(rand_arr)

[ 0.32392873  0.92563383  0.50651312  0.80914685  1.20735316 -1.82046066
  1.53800717  0.02111299  3.75719496 -0.55241365]


-   Accessing one element from an array

In [74]:
# accessing the 6 th entry of rand_arr
print(rand_arr[6])

1.53800716569196


-   Accessing multiple elements from an array

In [75]:
# we can access multiple entries at once using
print(rand_arr[4:9])

[ 1.20735316 -1.82046066  1.53800717  0.02111299  3.75719496]


In [76]:
# we can also access multiple non-consecutive entries using np.arange
print('Index of values to access: ',np.arange(3,10,3))
print(rand_arr[np.arange(3,10,3)])

Index of values to access:  [3 6 9]
[ 0.80914685  1.53800717 -0.55241365]


In [77]:
import numpy as np
matrix = np.array([[1,2,3],[4,5,6],[7,8,9]])
print(matrix[1][1])

5


In [78]:
import numpy as np
vec1 = np.array([4, 7, 8, 9, 10, 6, 1])
vec1[vec1>6] = 2
print(vec1)

[4 2 2 2 2 6 1]


**Accessing arrays using logical operations**

In [84]:
print(rand_arr)

[ 0.32392873  0.92563383  0.50651312  0.80914685  1.20735316 -1.82046066
  1.53800717  0.02111299  3.75719496 -0.55241365]


In [85]:
rand_arr > 0

array([ True,  True,  True,  True,  True, False,  True,  True,  True,
       False])

In [89]:
# accessing all the values of rand_arr which are greater than 0
print('Values greater than 0: ',rand_arr[rand_arr>0])

# accessing all the values of rand_arr which are less than 0
print('Values less than 0: ',rand_arr[rand_arr<0])

Values greater than 0:  [0.32392873 0.92563383 0.50651312 0.80914685 1.20735316 1.53800717
 0.02111299 3.75719496]
Values less than 0:  [-1.82046066 -0.55241365]


**Accessing the entries of a Matrix**

In [90]:
# let's generate an array with 10 random values
rand_mat = np.random.randn(5,5)
print(rand_mat)

[[-1.01817984 -0.52792623  0.42340045 -0.41820749  1.36130149]
 [ 1.83188664  0.62904945 -0.34187204 -0.29774022 -1.47655617]
 [-1.43749204 -0.71444986 -0.29301653 -0.80051094  0.15774212]
 [-0.49590769  0.59115983  1.02687445  0.74726049  1.82614095]
 [-0.49063728 -0.32911031 -0.81169518  0.71590593  0.28862913]]


In [91]:
# acessing the second row of the rand_mat
rand_mat[1]

array([ 1.83188664,  0.62904945, -0.34187204, -0.29774022, -1.47655617])

In [94]:
# acessing third element of the second row
print(rand_mat[1][2])

#or 
print(rand_mat[1,2])

-0.34187203695644314
-0.34187203695644314


In [96]:
# accessing first two rows with second and third column 
print(rand_mat[0:2,1:3])

[[-0.52792623  0.42340045]
 [ 0.62904945 -0.34187204]]


**Accessing matrices using logical operations**

In [ ]:
print(rand_mat)

In [ ]:
# accessing all the values of rand_mat which are greater than 0
print('Values greater than 0: \n ',rand_mat[rand_mat>0])

# accessing all the values of rand_mat which are less than 0
print('Values less than 0: \n',rand_mat[rand_mat<0])

**Modifying the entries of an Array**

In [97]:
print(rand_arr)

[ 0.32392873  0.92563383  0.50651312  0.80914685  1.20735316 -1.82046066
  1.53800717  0.02111299  3.75719496 -0.55241365]


In [ ]:
# let's change some values in an array!
# changing the values of index value 3 and index value 4 to 5
rand_arr[3:5] = 5
print(rand_arr)

In [ ]:
# changing the values of index value 0 and index value 1 to 2 and 3 respectively
rand_arr[0:2] = [2,3]
print(rand_arr)

In [ ]:
# modify entries using logical references
rand_arr[rand_arr>0] = 65
rand_arr

In [ ]:
import numpy as np
matrix = np.arange(1,10).reshape(3,3)
matrix[0:2,1:2] = 0
print(matrix)

**Modifying the entries of a Matrix**

In [ ]:
import numpy as np
array1 = np.array([[1,2,3],[4,5,6],[7,8,9]])
print(array1)
array2 = array1

In [ ]:
array2[1,2]= 100
print(array1)
print(array2)

In [ ]:
print(rand_mat3)

In [ ]:
# changing the values of the 4th and 5th element of the second and third rows of the matrix to 0
print('Matrix before modification: \n',rand_mat3)
rand_mat3[1:3,3:5] = 0
print('Matrix after modification: \n',rand_mat3)

In [ ]:
# extracting the first 2 rows and first 3 columns from the matrix
sub_mat = rand_mat[0:2,0:3]
print(sub_mat)

In [ ]:
# changing all the values of the extracted matrix to 3
sub_mat[:] = 3
print(sub_mat)

In [ ]:
# what happened to rand_mat when we change sub_mat?
rand_mat

In [ ]:
# to prevent this behavior we need to use the .copy() method when we assign sub_mat
# this behavior is the source of MANY errors for early python users!!!

rand_mat = np.random.randn(5,5)
print(rand_mat)
sub_mat = rand_mat[0:2,0:3].copy()
sub_mat[:] = 3
print(sub_mat)
print(rand_mat)

### 2.4 Saving and Loading a NumPy array<a href="#2.4-Saving-and-Loading-a-NumPy-array" class="anchor-link">¶</a>

**Let's save some NumPy objects on the disk for use later!**

In [98]:
from google.colab import drive
drive.mount('/content/drive')

ModuleNotFoundError: No module named 'google'

In [ ]:
# creating a random matrices
randint_matrix1 = np.random.randint(1,10,10).reshape(2,5)
print(randint_matrix1)
print('')
randint_matrix2 = np.random.randint(10,20,10).reshape(2,5)
print(randint_matrix2)

**Using np.save() function**


In [ ]:
np.save('randint_matrik',randint_matrix1)

**Using np.savez() function**

In [ ]:
np.savez('multiple_randint_matrik',randint_matrix1=randint_matrix1,randint_matrix2=randint_matrix2)

-   The files will be saved in the directory where the Jupyter Notebook
    is located.
-   With np.save() function, we can save an array/matrix to a NumPy .npy
    format.
-   np.savez() function has an advantage over np.save() function because
    with np.savez(), we can store several arrays/matrices into a single
    file in uncompressed .npz format.

In [ ]:
# now let's load it
loaded_arr = np.load('randint_matrik.npy')
loaded_multi = np.load('multiple_randint_matrik.npz')

print(loaded_arr)
print('')
print(loaded_multi)

-   We see that .npy file has been loaded but the .npz file is returning
    a memory location.
-   Let's see how to load the values stored in .npz file.

In [99]:
print('1st Matrix: \n',loaded_multi['randint_matrix1'])
print('2nd Matrix: \n',loaded_multi['randint_matrix2'])

new_matrix  = loaded_multi['randint_matrix1']
print('New Matrix: \n',new_matrix)

NameError: name 'loaded_multi' is not defined

In [ ]:
# we can also save/load text files...but only single variables
np.savetxt('text_file_name.txt',randint_matrix1,delimiter=',')
rand_mat_txt = np.loadtxt('text_file_name.txt',delimiter=',')
print(randint_matrix1)
print('')
print(rand_mat_txt)